`Frontc.ParseError` for `arm/_structs.h` on Apple M1

Most regression test fail due to some errors in include files.
First failing test:

```console
$ ./regtest.sh 00 07
./goblint --enable dbg.debug --enable dbg.showtemps --enable dbg.regression --sets warnstyle "legacy" --html   tests/regression/00-sanity/07-term2.c
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/mach/arm/_structs.h[443:0-0] : syntax error
Parsing errorFatal error: exception Frontc.ParseError("Parse error")
Raised at Frontc.parse_to_cabs_inner in file "src/frontc/frontc.ml", line 205, characters 6-39
Called from Frontc.parse_to_cabs in file "src/frontc/frontc.ml", line 124, characters 13-38
Called from Frontc.parse_helper in file "src/frontc/frontc.ml", line 256, characters 13-32
Called from Frontc.parse in file "src/frontc/frontc.ml" (inlined), line 264, characters 32-55
Called from Cilfacade.parse in file "src/util/cilfacade.ml", line 24, characters 2-26
Called from Stdlib__list.rev_map.rmap_f in file "list.ml", line 103, characters 22-25
Called from Maingoblint.merge_preprocessed in file "src/maingoblint.ml", line 256, characters 18-62
Called from Maingoblint.main in file "src/maingoblint.ml", line 460, characters 15-56
Called from Stdlib.at_exit.new_exit in file "stdlib.ml", line 554, characters 59-63
Called from Stdlib.do_at_exit in file "stdlib.ml" (inlined), line 560, characters 20-61
Called from Std_exit in file "std_exit.ml", line 18, characters 8-20
See result/index.xml
```

The syntax error is in `__uint128_t __v[32];` in
```c
#if defined(__arm64__)
_STRUCT_ARM_NEON_STATE64
{
	__uint128_t __v[32];
	__uint32_t  __fpsr;
	__uint32_t  __fpcr;
};

_STRUCT_ARM_NEON_STATE
{
	__uint128_t __v[16];
	__uint32_t  __fpsr;
	__uint32_t  __fpcr;
};
#elif defined(__arm__)
```

**The issue is that regardless of OS X or not, `__int128` is not supported.**
Cabs/FrontC can deal with it but anything CIL (i.e. starting with `cabs2cil.ml`) can not and he code converting the Cabs/FrontC type to a CIL type will go into the error case.

https://github.com/goblint/cil/blob/b2c66d3bdf09e2614d3b936b75eda5c5f1ce3587/src/frontc/cabs2cil.ml#L2639-L2640

https://github.com/goblint/cil/blob/b2c66d3bdf09e2614d3b936b75eda5c5f1ce3587/src/cil.ml#L281-L296

CIL also complains about this, but we have it set to continue on error, so it will generate boo inline assembly and dump some complaints to the logfile and then carry on.

~~~C
int main() {
  __int128 i;
  i = 0;

  int j = 24;
  j++;
}
~~~

```
./goblint 1.c --html
1.c:2: Error: Invalid combination of type specifiers
Error in doStatement (Errormsg.Error)
Live lines: 5
No lines with dead code found by solver (there might still be dead code removed by CIL).
Total lines (logical LoC): 5
Writing xml to temp. file: /tmp/ocaml15ad8ftmp
Time needed: 123 ms
1.c:3: Error: Cannot resolve variable i.
1.c:3: Error: Expected lval for assignment. Got Const(Int64(0,int,None))
```.

~~~C
int main(void) 
{ 
  int j ;

  {
  booo_statement: /* CIL Label */ 
#line 3
  __asm__  ("booo_exp(1.c:3)":);
#line 5
  j = 24;
#line 6
  j ++;
#line 7
  return (0);
}
~~~

`__asm__  ("booo_exp(1.c:3)":);` - what? 👻

This is all very confusing and fragile.
For your test above with a `typedef unsigned __int128 uint128_t;` and  `__uint128_t i;` I also get sensible errors:
```console
$ dune exec src/main.exe 1.c
1.c:1: Error: Invalid combination of type specifiers
Error on A.TYPEDEF (Errormsg.Error)
1.c:11: Error: Cannot find type uint128_t (kind:type)
Error in doStatement (Errormsg.Error)
Error: Cabs2cil had some errors
Fatal error: exception Errormsg.Error
Raised at Errormsg.s in file "src/ocamlutil/errormsg.ml", line 49, characters 17-28
Called from Dune__exe__Main.theMain in file "src/main.ml", line 219, characters 6-46
Called from Dune__exe__Main in file "src/main.ml", line 255, characters 4-14
```

If I use the type in the struct type declaration it's suddenly a `ParseError` without any information:
```console
$ cat 1.c
typedef unsigned __int128 uint128_t;

struct __darwin_arm_neon_state64
{
 __uint128_t __v[32];
 __uint32_t __fpsr;
 __uint32_t __fpcr;
};

int main() {
  // uint128_t i;
}

$ dune exec src/main.exe 1.c
1.c[5:0-0] : syntax error
Parsing errorFatal error: exception Frontc.ParseError("Parse error")
Raised at Errormsg.parse_error in file "src/ocamlutil/errormsg.ml", line 328, characters 2-27
Called from Stdlib__parsing.yyparse.loop in file "parsing.ml", line 151, characters 8-44
Called from Stdlib__parsing.yyparse in file "parsing.ml", line 164, characters 4-28
Re-raised at Stdlib__parsing.yyparse in file "parsing.ml", line 183, characters 8-17
Called from Stats.repeattime.repeatf in file "src/ocamlutil/stats.ml", line 106, characters 10-15
Re-raised at Stats.repeattime.repeatf in file "src/ocamlutil/stats.ml", line 110, characters 1-8
Called from Frontc.parse_to_cabs_inner in file "src/frontc/frontc.ml", line 191, characters 15-90
Re-raised at Frontc.parse_to_cabs_inner in file "src/frontc/frontc.ml", line 207, characters 6-73
Called from Frontc.parse_to_cabs in file "src/frontc/frontc.ml", line 124, characters 13-38
Called from Frontc.parse_helper in file "src/frontc/frontc.ml", line 258, characters 13-32
Called from Frontc.parse in file "src/frontc/frontc.ml", line 266, characters 32-55
Called from Dune__exe__Main.parseOneFile in file "src/main.ml", line 60, characters 12-28
Called from Util.count_map in file "src/ocamlutil/util.ml", line 19, characters 12-15
Called from Dune__exe__Main.theMain in file "src/main.ml", line 191, characters 16-64
Called from Dune__exe__Main in file "src/main.ml", line 255, characters 4-14
```

What should we do? Add proper support for `__int128`? Just define it as `long long` and don't care?
How often do we get those CIL Errors? Does it make sense to not ignore them?

--- a/src/cil.ml
+++ b/src/cil.ml
@@ -1312,13 +1312,15 @@ let isSigned = function
   | IUShort
   | IUInt
   | IULong
-  | IULongLong ->
+  | IULongLong
+  | IUInt128 ->
       false
   | ISChar
   | IShort
   | IInt
   | ILong
-  | ILongLong ->
+  | ILongLong
+  | IInt128 ->
       true
   | IChar ->
       not !M.theMachine.M.char_is_unsigned
@@ -1707,6 +1709,8 @@ let d_ikind () = function
   | IULongLong ->
       if !msvcMode then text "unsigned __int64"
       else text "unsigned long long"
+  | IInt128 -> text "__int128"
+  | IUInt128 -> text "unsigned __int128"
 
 let d_fkind () = function
     FFloat -> text "float"
@@ -1734,6 +1738,7 @@ let bytesSizeOfInt (ik: ikind): int =
   | IShort | IUShort -> !M.theMachine.M.sizeof_short
   | ILong | IULong -> !M.theMachine.M.sizeof_long
   | ILongLong | IULongLong -> !M.theMachine.M.sizeof_longlong
+  | IInt128 | IUInt128 -> 16
 
 (* constant *)
 let d_const () c =
@@ -1749,7 +1754,12 @@ let d_const () c =
         | IULong -> "UL"
         | ILongLong -> if !msvcMode then "L" else "LL"
         | IULongLong -> if !msvcMode then "UL" else "ULL"
+        (* if long long is 128 bit we can use its suffix, otherwise unsupported by GCC, see https://github.com/goblint/cil/issues/41#issuecomment-893291878 *)
+        | IInt128  when !M.theMachine.M.sizeof_longlong = 16 -> "LL"
+        | IUInt128 when !M.theMachine.M.sizeof_longlong = 16 -> "ULL"
         | _ -> ""
+          (* TODO warn/fail? *)
+          (* E.s (E.bug "unknown/unsupported suffix") *)
       in
       let prefix : string =
         if suffix <> "" then ""
@@ -2027,6 +2037,7 @@ let intRank (ik:ikind) : int =
   | IInt | IUInt -> 3
   | ILong | IULong -> 4
   | ILongLong | IULongLong -> 5
+  | IInt128 | IUInt128 -> 6 (* ? *)
 
 (* Return the common integer kind of the two integer arguments, as
    defined in ISO C 6.3.1.8 ("Usual arithmetic conversions") *)
@@ -2209,6 +2220,7 @@ let rec alignOf_int t =
     | TInt((IInt|IUInt), _) -> !M.theMachine.M.alignof_int
     | TInt((ILong|IULong), _) -> !M.theMachine.M.alignof_long
     | TInt((ILongLong|IULongLong), _) -> !M.theMachine.M.alignof_longlong
+    | TInt((IInt128|IUInt128), _) -> 16 (* ? *)
     | TEnum(ei, _) -> alignOf_int (TInt(ei.ekind, []))
     | TFloat(FFloat, _) -> !M.theMachine.M.alignof_float
     | TFloat(FDouble, _) -> !M.theMachine.M.alignof_double
@@ -7250,6 +7262,7 @@ let convertInts (i1:int64) (ik1:ikind) (i2:int64) (ik2:ikind)
       | IInt | IUInt -> 3
       | ILong | IULong -> 4
       | ILongLong | IULongLong -> 5
+      | IInt128 | IUInt128 -> 6 (* ? *)
     in
     let r1 = rank ik1 in
     let r2 = rank ik2 in

--- a/src/cil.ml
+++ b/src/cil.ml
@@ -2039,7 +2039,7 @@ let intRank (ik:ikind) : int =
   | IInt | IUInt -> 3
   | ILong | IULong -> 4
   | ILongLong | IULongLong -> 5
-  | IInt128 | IUInt128 -> 6 (* ? *)
+  | IInt128 | IUInt128 -> 6
 
 (* Return the common integer kind of the two integer arguments, as
    defined in ISO C 6.3.1.8 ("Usual arithmetic conversions") *)
@@ -2226,7 +2226,7 @@ let rec alignOf_int t =
     | TInt((IInt|IUInt), _) -> !M.theMachine.M.alignof_int
     | TInt((ILong|IULong), _) -> !M.theMachine.M.alignof_long
     | TInt((ILongLong|IULongLong), _) -> !M.theMachine.M.alignof_longlong
-    | TInt((IInt128|IUInt128), _) -> 16 (* ? *)
+    | TInt((IInt128|IUInt128), _) -> 16 (* not generated since not all architectures support 128bit ints and the value should be the same for those that do *)
     | TEnum(ei, _) -> alignOf_int (TInt(ei.ekind, []))
     | TFloat(FFloat, _) -> !M.theMachine.M.alignof_float
     | TFloat(FDouble, _) -> !M.theMachine.M.alignof_double
@@ -2859,9 +2859,9 @@ let parseInt (str: string) : exp =
   (* The length of the suffix and a list of possible kinds. See ISO
   * 6.4.4.1 *)
   let hasSuffix = hasSuffix str in
-  let suffixlen, kinds = (* TODO check situation with __int128 *)
+  let suffixlen, kinds = (* 128bit constants are only supported if long long is also 128bit, so we can parse those as long long *)
     if hasSuffix "ULL" || hasSuffix "LLU" then
-      3, [IULongLong] (* ? [if !M.theMachine.M.sizeof_longlong = 16 then IUInt128 else IULongLong] *)
+      3, [IULongLong]
     else if hasSuffix "LL" then
       2, if octalhex then [ILongLong; IULongLong] else [ILongLong]
     else if hasSuffix "UL" || hasSuffix "LU" then
@@ -7270,7 +7270,7 @@ let convertInts (i1:int64) (ik1:ikind) (i2:int64) (ik2:ikind)
       | IInt | IUInt -> 3
       | ILong | IULong -> 4
       | ILongLong | IULongLong -> 5
-      | IInt128 | IUInt128 -> 6 (* ? *)
+      | IInt128 | IUInt128 -> 6
     in
     let r1 = rank ik1 in
     let r2 = rank ik2 in

--- a/src/cil.ml
+++ b/src/cil.ml
@@ -293,6 +293,8 @@ and ikind =
   | ILongLong   (** [long long] (or [_int64] on Microsoft Visual C) *)
   | IULongLong  (** [unsigned long long] (or [unsigned _int64] on Microsoft
                     Visual C) *)
+  | IInt128     (** [__int128] *)
+  | IUInt128    (** [unsigned __int128] *)
 
 (** Various kinds of floating-point numbers*)
 and fkind =

--- a/src/cil.ml
+++ b/src/cil.ml
@@ -2017,6 +2017,7 @@ let unsignedVersionOf (ik:ikind): ikind =
   | IInt -> IUInt
   | ILong -> IULong
   | ILongLong -> IULongLong
+  | IInt128 -> IUInt128
   | _ -> ik
 
 let signedVersionOf (ik:ikind): ikind =
@@ -2026,6 +2027,7 @@ let signedVersionOf (ik:ikind): ikind =
   | IUInt -> IInt
   | IULong -> ILong
   | IULongLong -> ILongLong
+  | IUInt128 -> IInt128
   | _ -> ik
 
 (* Return the integer conversion rank of an integer kind *)
@@ -2070,6 +2072,7 @@ let intKindForSize (s:int) (unsigned:bool) : ikind =
     else if s = !M.theMachine.M.sizeof_long then IULong
     else if s = !M.theMachine.M.sizeof_short then IUShort
     else if s = !M.theMachine.M.sizeof_longlong then IULongLong
+    else if s = 16 then IUInt128
     else raise Not_found
   else
     (* Test the most common sizes first *)
@@ -2078,6 +2081,7 @@ let intKindForSize (s:int) (unsigned:bool) : ikind =
     else if s = !M.theMachine.M.sizeof_long then ILong
     else if s = !M.theMachine.M.sizeof_short then IShort
     else if s = !M.theMachine.M.sizeof_longlong then ILongLong
+    else if s = 16 then IInt128
     else raise Not_found
 
 let floatKindForSize (s:int) =
@@ -2155,13 +2159,15 @@ let intKindForValue (i: cilint) (unsigned: bool) =
     else if fitsInInt IUShort i then IUShort
     else if fitsInInt IUInt i then IUInt
     else if fitsInInt IULong i then IULong
-    else IULongLong
+    else if fitsInInt IUInt128 i then IUInt128
+    else IULongLong (* warn, IUInt128? *)
   else
     if fitsInInt ISChar i then ISChar
     else if fitsInInt IShort i then IShort
     else if fitsInInt IInt i then IInt
     else if fitsInInt ILong i then ILong
-    else ILongLong
+    else if fitsInInt IInt128 i then IInt128
+    else ILongLong (* warn, IInt128? *)
 
 (** If the given expression is an integer constant or a CastE'd
     integer constant, return that constant's value as an ikind, int64 pair.
@@ -2853,9 +2859,9 @@ let parseInt (str: string) : exp =
   (* The length of the suffix and a list of possible kinds. See ISO
   * 6.4.4.1 *)
   let hasSuffix = hasSuffix str in
-  let suffixlen, kinds =
+  let suffixlen, kinds = (* TODO check situation with __int128 *)
     if hasSuffix "ULL" || hasSuffix "LLU" then
-      3, [IULongLong]
+      3, [IULongLong] (* ? [if !M.theMachine.M.sizeof_longlong = 16 then IUInt128 else IULongLong] *)
     else if hasSuffix "LL" then
       2, if octalhex then [ILongLong; IULongLong] else [ILongLong]
     else if hasSuffix "UL" || hasSuffix "LU" then
@@ -7039,6 +7045,8 @@ let initCIL () =
       else if name = "unsigned short" then IUShort
       else if name = "char" then IChar
       else if name = "unsigned char" then IUChar
+      else if name = "__int128" then IInt128
+      else if name = "unsigned __int128" then IUInt128
       else E.s(E.unimp "initCIL: cannot find the right ikind for type %s\n" name)
     in
     upointType := TInt(findIkindSz true !M.theMachine.M.sizeof_ptr, []);

--- a/src/cil.mli
+++ b/src/cil.mli
@@ -287,6 +287,8 @@ and ikind =
   | ILongLong   (** [long long] (or [_int64] on Microsoft Visual C) *)
   | IULongLong  (** [unsigned long long] (or [unsigned _int64] on Microsoft
                     Visual C) *)
+  | IInt128     (** [__int128] *)
+  | IUInt128    (** [unsigned __int128] *)
 
 (** Various kinds of floating-point numbers*)
 and fkind =

--- a/src/ciltools.ml
+++ b/src/ciltools.ml
@@ -59,6 +59,7 @@ let ocaml_int_to_cil v n s =
   let short_size = bitsSizeOf (TInt(IShort,[]))in
   let long_size = bitsSizeOf longType in
   let longlong_size = bitsSizeOf (TInt(ILongLong,[])) in
+  let int128_size = bitsSizeOf (TInt(IInt128,[])) in
   let i =
     match s with
       Signed ->
@@ -72,6 +73,8 @@ let ocaml_int_to_cil v n s =
 	  ILong
 	else if (n = longlong_size) then
 	  ILongLong
+	else if (n = int128_size) then
+	  IInt128
 	else
 	  raise Weird_bitwidth
     | Unsigned ->
@@ -85,6 +88,8 @@ let ocaml_int_to_cil v n s =
 	  IULong
 	else if (n = longlong_size) then
 	  IULongLong
+	else if (n = int128_size) then
+	  IUInt128
 	else
 	  raise Weird_bitwidth
   in

--- a/src/formatparse.mly
+++ b/src/formatparse.mly
@@ -238,7 +238,7 @@ type maybeInit =
 %token EOF
 %token CHAR INT DOUBLE FLOAT VOID INT64 INT32
 %token ENUM STRUCT TYPEDEF UNION
-%token SIGNED UNSIGNED LONG SHORT
+%token SIGNED UNSIGNED LONG SHORT INT128
 %token VOLATILE EXTERN STATIC CONST RESTRICT AUTO REGISTER
 
 %token <string> ARG_e ARG_eo ARG_E ARG_u ARG_b ARG_t ARG_d ARG_lo ARG_l ARG_i
@@ -815,6 +815,15 @@ type_spec:
 |   UNSIGNED LONG LONG    { ((fun al args -> TInt(IULongLong, al)),
 
                              matchIntType IULongLong)
+                           }
+
+|   INT128          { ((fun al args -> TInt(IInt128, al)),
+
+                          matchIntType IInt128)
+                        }
+|   UNSIGNED INT128    { ((fun al args -> TInt(IUInt128, al)),
+
+                             matchIntType IUInt128)
                            }
 
 |   FLOAT           { ((fun al args -> TFloat(FFloat, al)),

--- a/src/frontc/cabs2cil.ml
+++ b/src/frontc/cabs2cil.ml
@@ -250,7 +250,7 @@ let interpret_character_constant char_list =
     if value <= (Int64.of_int32 Int32.max_int) then
       (CInt64(value,IULong,orig_rep)),(TInt(IULong,[]))
     else
-      (CInt64(value,IULongLong,orig_rep)),(TInt(IULongLong,[])) (* TODO no need to handle __int128? *)
+      (CInt64(value,IULongLong,orig_rep)),(TInt(IULongLong,[])) (* 128bit constants are only supported if long long is also 128bit wide *)
   end
 
 (*** EXPRESSIONS *************)
@@ -2554,9 +2554,7 @@ let rec doSpecList (suggestedAnonName: string) (* This string will be part of
 	    else if fitsInInt ILong i then ILong
 	    else if fitsInInt IULong i then IULong
 	    else if fitsInInt ILongLong i then ILongLong
-	    else if fitsInInt IInt128 i then IInt128
-	    else if fitsInInt IUInt128 i then IUInt128
-	    else IULongLong (* TODO warn, use IUInt128? *)
+	    else IULongLong (* assume there can be not enum constants that don't fit in long long since there can only be 128bit constants if long long is also 128bit *)
 	in
         (* as each name,value pair is determined, this is called *)
         let rec processName kname (i: exp) loc rest = begin

--- a/src/frontc/cabs2cil.ml
+++ b/src/frontc/cabs2cil.ml
@@ -743,11 +743,11 @@ let lookupTypeNoError (kind: string)
   | _ -> raise Not_found
 
 let lookupType (kind: string)
-               (n: string) : typ * location =
+               (n: string) : (typ * location) option =
   try
-    lookupTypeNoError kind n
-  with Not_found ->
-    E.s (error "Cannot find type %s (kind:%s)" n kind)
+    Some (lookupTypeNoError kind n)
+  with Not_found -> None
+    (* E.s (error "Cannot find type %s (kind:%s)" n kind) *)
 
 (* Create the self ref cell and add it to the map. Return also an indication
  * if this is a new one. *)
@@ -2471,8 +2471,12 @@ let rec doSpecList (suggestedAnonName: string) (* This string will be part of
         end else
           let t =
             match lookupType "type" n with
-              (TNamed _) as x, _ -> x
-            | typ -> E.s (error "Named type %s is not mapped correctly" n)
+            | Some (TNamed _ as x, _) -> x
+            | _ ->
+              match n with
+              | "__int128_t" -> TInt(IInt128, [])
+              | "__uint128_t" -> TInt(IUInt128, [])
+              | _ -> E.s (error "Named type %s is not mapped correctly" n)
           in
           t
     end

--- a/src/frontc/cabs2cil.ml
+++ b/src/frontc/cabs2cil.ml
@@ -250,7 +250,7 @@ let interpret_character_constant char_list =
     if value <= (Int64.of_int32 Int32.max_int) then
       (CInt64(value,IULong,orig_rep)),(TInt(IULong,[]))
     else
-      (CInt64(value,IULongLong,orig_rep)),(TInt(IULongLong,[]))
+      (CInt64(value,IULongLong,orig_rep)),(TInt(IULongLong,[])) (* TODO no need to handle __int128? *)
   end
 
 (*** EXPRESSIONS *************)
@@ -2554,7 +2554,9 @@ let rec doSpecList (suggestedAnonName: string) (* This string will be part of
 	    else if fitsInInt ILong i then ILong
 	    else if fitsInInt IULong i then IULong
 	    else if fitsInInt ILongLong i then ILongLong
-	    else IULongLong
+	    else if fitsInInt IInt128 i then IInt128
+	    else if fitsInInt IUInt128 i then IUInt128
+	    else IULongLong (* TODO warn, use IUInt128? *)
 	in
         (* as each name,value pair is determined, this is called *)
         let rec processName kname (i: exp) loc rest = begin

--- a/src/frontc/cabs2cil.ml
+++ b/src/frontc/cabs2cil.ml
@@ -2382,9 +2382,10 @@ let rec doSpecList (suggestedAnonName: string) (* This string will be part of
       | A.Tlong -> 5
       | A.Tint -> 6
       | A.Tint64 -> 7
-      | A.Tfloat -> 8
-      | A.Tdouble -> 9
-      | _ -> 10 (* There should be at most one of the others *)
+      | A.Tint128 -> 8
+      | A.Tfloat -> 9
+      | A.Tdouble -> 10
+      | _ -> 11 (* There should be at most one of the others *)
     in
     List.stable_sort (fun ts1 ts2 -> compare (order ts1) (order ts2)) tspecs'
   in
@@ -2450,6 +2451,11 @@ let rec doSpecList (suggestedAnonName: string) (* This string will be part of
     | [A.Tsigned; A.Tint64] -> TInt(ILongLong, [])
 
     | [A.Tunsigned; A.Tint64] -> TInt(IULongLong, [])
+
+    | [A.Tint128]
+    | [A.Tsigned; A.Tint128] -> TInt(IInt128, [])
+
+    | [A.Tunsigned; A.Tint128] -> TInt(IUInt128, [])
 
     | [A.Tfloat] -> TFloat(FFloat, [])
     | [A.Tdouble] -> TFloat(FDouble, [])

--- a/src/frontc/clexer.mll
+++ b/src/frontc/clexer.mll
@@ -279,6 +279,9 @@ let init ~(filename: string) : Lexing.lexbuf =
   init_lexicon ();
   (* Inititialize the pointer in Errormsg *)
   Lexerhack.add_type := add_type;
+  (* add some built-in types which are handled as Tnamed in cabs2cil *)
+  add_type "__int128_t"; (* __int128 *)
+  add_type "__uint128_t"; (* unsigned __int128 *)
   Lexerhack.push_context := push_context;
   Lexerhack.pop_context := pop_context;
   Lexerhack.add_identifier := add_identifier;

--- a/src/frontc/frontc.ml
+++ b/src/frontc/frontc.ml
@@ -187,6 +187,7 @@ and parse_to_cabs_inner (fname : string) =
   try
     if !E.verboseFlag then ignore (E.log "Frontc is parsing %s\n" fname);
     flush !E.logChannel;
+    (* if !E.verboseFlag then ignore @@ Parsing.set_trace true; *)
     let lexbuf = Clexer.init ~filename:fname in
     let cabs = Stats.time "parse" (Cparser.interpret (Whitetrack.wraplexer clexer)) lexbuf in
     Whitetrack.setFinalWhite (Clexer.get_white ());

--- a/src/frontc/frontc.ml
+++ b/src/frontc/frontc.ml
@@ -202,7 +202,9 @@ and parse_to_cabs_inner (fname : string) =
       ignore (E.log "Parsing error");
       Clexer.finish ();
       close_output ();
-      raise (ParseError("Parse error"))
+      (* raise (ParseError("Parse error")) *)
+      let backtrace = Printexc.get_raw_backtrace () in
+      Printexc.raise_with_backtrace (ParseError("Parse error")) backtrace (* re-raise with captured inner backtrace *)
   end
   | e -> begin
       ignore (E.log "Caught %s while parsing\n" (Printexc.to_string e));

